lib/body.ml

```diff
@@ -144,10 +144,13 @@ let close_reader t =
 ;;
 
 let when_ready_to_write t callback =
-  if is_closed t then callback ();
-  if not (t.when_ready_to_write == default_ready_to_write)
-  then failwith "Body.when_ready_to_write: only one callback can be registered at a time";
-  t.when_ready_to_write <- callback
+  if is_closed t
+  then callback ()
+  else begin
+    if not (t.when_ready_to_write == default_ready_to_write)
+    then failwith "Body.when_ready_to_write: only one callback can be registered at a time";
```

Comment:
Now that I'm looking at this again, it seems like the check to see if another callback's registered should precede the other logic in this function.

Reply:
Yeah, excellent point. I changed that too.